src/Aspire.Hosting.Azure.ServiceBus/AzureServiceBusExtensions.cs

```diff
@@ -124,19 +151,381 @@ public static IResourceBuilder<AzureServiceBusResource> AddQueue(this IResourceB
+        string[]? topicNames = null;
+
+        builder.ApplicationBuilder.Eventing.Subscribe<ConnectionStringAvailableEvent>(builder.Resource, async (@event, ct) =>
+        {
+            var serviceBusEmulatorResources = builder.ApplicationBuilder.Resources.OfType<AzureServiceBusResource>().Where(x => x is { } serviceBusResource && serviceBusResource.IsEmulator);
+
+            if (!serviceBusEmulatorResources.Any())
+            {
+                // No-op if there is no Azure Service Bus emulator resource.
+                return;
+            }
+
+            connectionString = await builder.Resource.ConnectionStringExpression.GetValueAsync(ct).ConfigureAwait(false);
+
+            if (connectionString == null)
+            {
+                throw new DistributedApplicationException($"ConnectionStringAvailableEvent was published for the '{builder.Resource.Name}' resource but the connection string was null.");
+            }
+
+            queueNames = serviceBusEmulatorResources.SelectMany(x => x.Queues).Select(x => x.Name).ToArray();
+            topicNames = serviceBusEmulatorResources.SelectMany(x => x.Topics).Select(x => x.Name).ToArray();
+        });
+
+        var healthCheckKey = $"{builder.Resource.Name}_check";
+
+        if (configureContainer != null)
+        {
+            var surrogate = new AzureServiceBusEmulatorResource(builder.Resource);
+            var surrogateBuilder = builder.ApplicationBuilder.CreateResourceBuilder(surrogate);
+            configureContainer(surrogateBuilder);
+        }
+
+        // To use the existing ServiceBus health check we would need to know if there is any queue or topic defined.
+        // We can register a health check for a queue and then no-op if there are no queues. Same for topics. This
+        // custom ServiceBusHealthCheck can be registered and will then iterate over the queues and topics.
+        // If no queues or no topics are defined then the health check will be successful.
+
+        var noRetryOptions = new ServiceBusClientOptions { RetryOptions = new ServiceBusRetryOptions { MaxRetries = 0 } };
+
+        builder.ApplicationBuilder.Services.AddHealthChecks()
+          .Add(new HealthCheckRegistration(
+              healthCheckKey,
+              sp => new ServiceBusHealthCheck(
+                  () => new ServiceBusClient(connectionString, noRetryOptions) ?? throw new DistributedApplicationException($"{nameof(connectionString)} was not initialized."),
+                  () => queueNames ?? throw new DistributedApplicationException($"{nameof(queueNames)} was not initialized."),
+                  () => topicNames ?? throw new DistributedApplicationException($"{nameof(topicNames)} was not initialized.")),
+              failureStatus: default,
+              tags: default,
+              timeout: default));
+
+        builder.WithHealthCheck(healthCheckKey);
+
+        builder.ApplicationBuilder.Eventing.Subscribe<AfterEndpointsAllocatedEvent>((e, ct) =>
```

Comment:
Does this need to be 2 separate event subscriptions? Can we just do everything (get the connection string and queue names plus write the json file) at one time?